### PR TITLE
Stashing checkpointing RNG states based on devices of arg tensors

### PR DESCRIPTION
**Pull request by mcarilli**
This PR intends to address @apaszke's concerns in https://github.com/pytorch/pytorch/pull/14253#issuecomment-441740016.  Preserving the rng state is now controlled by a kwarg rather than a global state, hopefully in a python 2.7-compatible way.

Additionally, the checkpointing function stashes and restores the RNG states of 
1. devices associated with all input tensor args to run_fn as well as 
2. the current device.

I could easily change this to only save and restore the RNG states associated 1. alone.  This would simplify the logic to create a [deduplicated, ordered](https://github.com/pytorch/pytorch/compare/master...mcarilli:checkpointing_rng_touchup?expand=1#diff-58da227fc9b1d56752b7dfad90428fe0R37) list of devices considered active.

I'm wondering if the [get_device_states](https://github.com/pytorch/pytorch/compare/master...mcarilli:checkpointing_rng_touchup?expand=1#diff-58da227fc9b1d56752b7dfad90428fe0R32) and [set_device_states](https://github.com/pytorch/pytorch/compare/master...mcarilli:checkpointing_rng_touchup?expand=1#diff-58da227fc9b1d56752b7dfad90428fe0R47) functions are general enough to reside elsewhere (presumably torch/random.py).  I'm also wondering if the check on [torch.cuda._initialized](https://github.com/pytorch/pytorch/compare/master...mcarilli:checkpointing_rng_touchup?expand=1#diff-58da227fc9b1d56752b7dfad90428fe0R47) would be better placed within `get_device_states`.

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -21,38 +21,52 @@ def check_backward_validity(inputs):
...
 
-# Global switch to toggle whether or not checkpointed passes stash and restore
-# the RNG state.  If True, any checkpoints making use of RNG should achieve deterministic
-# output compared to non-checkpointed passes.
-preserve_rng_state = True
+# We can't know if the run_fn will internally move some args to a device
+# other than the current device, which would require logic to preserve
+# rng states for those devices as well.  We could paranoically stash and restore
+# ALL the rng states for all visible devices, but that seems very wasteful for
+# most cases.  Compromise:  Stash the RNG state for the current device as well as
```
Comment (apaszke) [inline torch/utils/checkpoint.py]: Nooo, there's no point in stashing the values on the current device I think. What this compromise is, is an assumption that you will only do compute on the devices of your inputs. We should probably just stash for all devices you can find among them, and only those

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (mcarilli) [reply torch/utils/checkpoint.py]: Ok, changed in latest commit

**Comment by mcarilli** [comment]
@apaszke Test failures appear unrelated.

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -66,23 +78,25 @@ def backward(ctx, *args):
         # Stash the surrounding rng state, and mimic the state that was
         # present at this time during forward.  Restore the surrouding state
         # when we're done.
-        rng_devices = [torch.cuda.current_device()] if ctx.had_cuda_in_fwd else []
-        with torch.random.fork_rng(devices=rng_devices, enabled=preserve_rng_state):
-            if preserve_rng_state:
-                torch.set_rng_state(ctx.fwd_cpu_rng_state)
+        rng_devices = []
+        if ctx.preserve_rng_state and ctx.had_cuda_in_fwd:
+            rng_devices = ctx.fwd_gpu_devices
```
Comment (apaszke) [inline torch/utils/checkpoint.py]: Well, `had_cuda_in_fwd` shouldn't be a prerequisite here, because we still should restore the CPU state.

**Review by apaszke — COMMENTED** [review]
Looks good. Only the logic will not do anything when the function runs on CPU, which is a mistake.

File: torch/utils/checkpoint.py (hunk 2 above)
Comment (mcarilli) [reply torch/utils/checkpoint.py]: I'm not quite sure what you mean.  `rng_devices` only controls the GPU devices for which the state should be saved and restored by the context manager.  If cuda was not used in forward, `rng_devices` should be `[]` (which is how torch.random.fork_rng expects to be told "you don't need to mess with any GPU RNG states").

The torch.random.fork_rng context manager [always saves and restores the CPU state unconditionally](https://github.com/pytorch/pytorch/blob/master/torch/random.py#L108), even if the `devices` kwarg is `[]`.

Comment (apaszke) [reply torch/utils/checkpoint.py]: Ohh I see, makes sense. Thanks!

**Review by apaszke — APPROVED** [review]

**Pull request closed without merging**